File: ios/Packages/MeshtasticKit/Sources/MeshtasticKit/Protocol/MeshtasticProto.swift
//
//  MeshtasticProto.swift
//  MeshtasticKit
//
//  The Meshtastic client protobuf conversation, phone side.
//

import Foundation

/// The wire format a Lilyshark T-Deck already answers over BLE.
///
/// After connecting, the phone writes `ToRadio{want_config_id: nonce}` and then
/// reads `FromRadio` until `config_complete_id` echoes that nonce; what arrives
/// in between is everything the app will ever show about the device. Afterwards
/// the two ends simply trade packets.
///
/// Field numbers come from meshtastic/protobufs master (mesh.proto,
/// deviceonly.proto), and every byte here is pinned by vectors that the
/// firmware's `test/meshtastic_api` and the web client's `meshtasticProto.test`
/// pin as well. The three implementations cannot drift apart without one of the
/// three suites going red, which is the point of hand-rolling a third copy
/// rather than depending on a generated one.
///
/// Pure functions over `Data`: nothing here touches CoreBluetooth, so all of it
/// runs under `swift test` with no radio in the room.
public enum MeshtasticProto {
    /// The address every node listens on.
    public static let broadcast: UInt32 = 0xffff_ffff

    private static let portText: UInt64 = 1
    private static let portPosition: UInt64 = 3
    private static let portRouting: UInt64 = 5

    // MARK: - ToRadio (phone -> deck)

    /// `ToRadio{want_config_id}`, the message that starts the config dump.
    public static func encodeWantConfig(nonce: UInt32) -> Data {
        var writer = ProtoWriter()
        // Written even when the nonce is zero. Everywhere else a zero field is
        // omitted, but the deck echoes this one back to mark the dump finished,
        // and a phone that omitted it would wait forever for an echo of nothing.
        writer.tag(3, ProtoWriter.wireVarint)
        writer.varint(nonce)
        return Data(writer.bytes)
    }

    /// `ToRadio{packet}` carrying a text message for the mesh.
    ///
    /// Pass `broadcast` as `to` for a channel message, or a node number for a
    /// direct message. With `wantAck` set the deck answers with a Routing
    /// result naming `packetID`, which is how the app learns the message left.
    public static func encodeTextPacket(
        to: UInt32,
        channel: UInt32,
        packetID: UInt32,
        text: String,
        wantAck: Bool
    ) -> Data {
        var data = ProtoWriter()
        data.uint(1, UInt32(portText))
        data.bytesField(2, Array(text.utf8))

        var packet = ProtoWriter()
        packet.fixed32(2, to)
        packet.uint(3, channel)
        packet.message(4, data)
        packet.fixed32(6, packetID)
        packet.uint(10, wantAck ? 1 : 0)

        var toRadio = ProtoWriter()
        toRadio.message(1, packet)
        return Data(toRadio.bytes)
    }

    // MARK: - FromRadio (deck -> phone)

    /// One node as the deck describes it, in the config dump or on arrival.
    public struct NodeInfo: Equatable, Sendable {
        public var num: UInt32
        public var id: String
        public var longName: String
        public var shortName: String
        public var snr: Float?
        public var latitude: Double?
        public var longitude: Double?

        public init(
            num: UInt32,
            id: String,
            longName: String,
            shortName: String,
            snr: Float? = nil,
            latitude: Double? = nil,
            longitude: Double? = nil
        ) {
            self.num = num
            self.id = id
            self.longName = longName
            self.shortName = shortName
            self.snr = snr
            self.latitude = latitude
            self.longitude = longitude
        }
    }

    /// A text heard on the mesh. `rxSNR` and `rxRSSI` are absent for a message
    /// the deck itself originated, since nothing received it over the air.
    public struct TextMessage: Equatable, Sendable {
        public var from: UInt32
        public var to: UInt32
        public var packetID: UInt32
        public var channel: UInt32
        public var text: String
        public var rxSNR: Float?
        public var rxRSSI: Int32?

        public init(
            from: UInt32,
            to: UInt32,
            packetID: UInt32,
            channel: UInt32,
            text: String,
            rxSNR: Float? = nil,
            rxRSSI: Int32? = nil
        ) {
            self.from = from
            self.to = to
            self.packetID = packetID
            self.channel = channel
            self.text = text
            self.rxSNR = rxSNR
            self.rxRSSI = rxRSSI
        }
    }

    /// A position heard on the mesh, already converted out of Meshtastic's
    /// 1e-7-degree integers into degrees.
    public struct Position: Equatable, Sendable {
        public var from: UInt32
        public var latitude: Double
        public var longitude: Double

        public init(from: UInt32, latitude: Double, longitude: Double) {
            self.from = from
            self.latitude = latitude
            self.longitude = longitude
        }
    }

    /// Everything a phone acts on, and `other` for everything it does not.
    ///
    /// `other` is not a failure: the radio sends channel and config messages
    /// this client has no use for, and they still have to be read past to reach
    /// `configComplete`.
    public enum FromRadio: Equatable, Sendable {
        case myInfo(num: UInt32)
        case metadata(firmware: String)
        case nodeInfo(NodeInfo)
        case configComplete(nonce: UInt32)
        case text(TextMessage)
        case position(Position)
        case routing(requestID: UInt32, error: UInt32)
        case other
    }

    /// Parse one `FromRadio` protobuf, or `nil` if the bytes are malformed.
    ///
    /// The distinction matters on this transport: an unreadable value means the
    /// link is confused and the connection should be torn down, whereas a
    /// well-formed message this client ignores is ordinary traffic.
    public static func parseFromRadio(_ data: Data) -> FromRadio? {
        guard let fields = readFields(Array(data)) else { return nil }
        for field in fields {
            switch field.number {
            case 2:
                if let bytes = field.bytes { return parsePacket(bytes) }
            case 3:
                guard let bytes = field.bytes else { break }
                for entry in readFields(bytes) ?? []
                where entry.number == 1 && entry.wireType == ProtoWriter.wireVarint {
                    return .myInfo(num: truncate(entry.value))
                }
                return .myInfo(num: 0)
            case 4:
                if let bytes = field.bytes { return parseNodeInfo(bytes) }
            case 7:
                return .configComplete(nonce: truncate(field.value))
            case 13:
                guard let bytes = field.bytes else { break }
                for entry in readFields(bytes) ?? [] where entry.number == 1 {
                    if let firmware = entry.bytes {
                        return .metadata(firmware: string(firmware))
                    }
                }
                return .metadata(firmware: "")
            default:
                break
            }
        }
        return .other
    }

    private static func parseNodeInfo(_ bytes: [UInt8]) -> FromRadio {
        var info = NodeInfo(num: 0, id: "", longName: "", shortName: "")
        for field in readFields(bytes) ?? [] {
            switch (field.number, field.wireType) {
            case (1, ProtoWriter.wireVarint):
                info.num = truncate(field.value)
            case (2, ProtoWriter.wireLength):
                for entry in readFields(field.bytes ?? []) ?? [] {
                    guard let text = entry.bytes else { continue }
                    switch entry.number {
                    case 1: info.id = string(text)
                    case 2: info.longName = string(text)
                    case 3: info.shortName = string(text)
                    default: break
                    }
                }
            case (3, ProtoWriter.wireLength):
                for entry in readFields(field.bytes ?? []) ?? [] {
                    guard entry.wireType == ProtoWriter.wireFixed32 else { continue }
                    if entry.number == 1 { info.latitude = degrees(entry.value) }
                    if entry.number == 2 { info.longitude = degrees(entry.value) }
                }
            case (4, ProtoWriter.wireFixed32):
                info.snr = Float(bitPattern: truncate(field.value))
            default:
                break
            }
        }
        return .nodeInfo(info)
    }

    private static func parsePacket(_ bytes: [UInt8]) -> FromRadio {
        var from: UInt32 = 0
        var to: UInt32 = 0
        var packetID: UInt32 = 0
        var channel: UInt32 = 0
        var rxSNR: Float?
        var rxRSSI: Int32?
        var decoded: [UInt8]?
        for field in readFields(bytes) ?? [] {
            switch (field.number, field.wireType) {
            case (1, ProtoWriter.wireFixed32): from = truncate(field.value)
            case (2, ProtoWriter.wireFixed32): to = truncate(field.value)
            case (3, ProtoWriter.wireVarint): channel = truncate(field.value)
            case (4, ProtoWriter.wireLength): decoded = field.bytes
            case (6, ProtoWriter.wireFixed32): packetID = truncate(field.value)
            case (8, ProtoWriter.wireFixed32): rxSNR = Float(bitPattern: truncate(field.value))
            case (12, ProtoWriter.wireVarint):
                // rx_rssi is a protobuf int32, so a negative one arrives as a
                // ten-byte varint sign-extended to 64 bits -- which is exactly
                // what the firmware's field_int32 writes. Reading the low 32
                // bits back as signed recovers it, and also handles the
                // truncated five-byte form other implementations emit.
                rxRSSI = Int32(truncatingIfNeeded: field.value)
            default:
                break
            }
        }

        guard let decoded else { return .other }
        var portnum: UInt64 = 0
        var payload: [UInt8] = []
        var requestID: UInt32 = 0
        for field in readFields(decoded) ?? [] {
            switch (field.number, field.wireType) {
            case (1, ProtoWriter.wireVarint): portnum = field.value
            case (2, ProtoWriter.wireLength): payload = field.bytes ?? []
            case (6, ProtoWriter.wireFixed32): requestID = truncate(field.value)
            default: break
            }
        }

        switch portnum {
        case portText:
            return .text(
                TextMessage(
                    from: from,
                    to: to,
                    packetID: packetID,
                    channel: channel,
                    text: string(payload),
                    rxSNR: rxSNR,
                    rxRSSI: rxRSSI
                )
            )
        case portPosition:
            var latitude = 0.0
            var longitude = 0.0
            for field in readFields(payload) ?? []
            where field.wireType == ProtoWriter.wireFixed32 {
                if field.number == 1 { latitude = degrees(field.value) }
                if field.number == 2 { longitude = degrees(field.value) }
            }
            return .position(Position(from: from, latitude: latitude, longitude: longitude))
        case portRouting:
            var error: UInt32 = 0
            for field in readFields(payload) ?? []
            where field.number == 3 && field.wireType == ProtoWriter.wireVarint {
                error = truncate(field.value)
            }
            return .routing(requestID: requestID, error: error)
        default:
            return .other
        }
    }

    // MARK: - Conversions

    private static func truncate(_ value: UInt64) -> UInt32 {
        UInt32(truncatingIfNeeded: value)
    }

    /// Meshtastic carries coordinates as signed 1e-7 degrees in an sfixed32.
    private static func degrees(_ bits: UInt64) -> Double {
        Double(Int32(bitPattern: truncate(bits))) / 1e7
    }

    /// Names and message bodies are UTF-8. Anything that is not becomes
    /// replacement characters rather than a parse failure, because a neighbour
    /// with a mojibake name is still a neighbour worth showing.
    private static func string(_ bytes: [UInt8]) -> String {
        String(decoding: bytes, as: UTF8.self)
    }
}

// MARK: - Writer

/// The little of protobuf this client emits, written by hand.
///
/// Only the wire types Meshtastic uses for the two messages we send are here.
/// Following the reference implementations, a zero-valued field is omitted --
/// protobuf's default -- with `encodeWantConfig` the single documented
/// exception.
private struct ProtoWriter {
    static let wireVarint = 0
    static let wireFixed32 = 5
    static let wireLength = 2

    private(set) var bytes: [UInt8] = []

    mutating func varint(_ value: UInt32) {
        var remaining = value
        repeat {
            let part = UInt8(remaining & 0x7f)
            remaining >>= 7
            bytes.append(remaining != 0 ? part | 0x80 : part)
        } while remaining != 0
    }

    mutating func tag(_ field: UInt32, _ wireType: Int) {
        varint((field << 3) | UInt32(wireType))
    }

    mutating func uint(_ field: UInt32, _ value: UInt32) {
        guard value != 0 else { return }
        tag(field, Self.wireVarint)
        varint(value)
    }

    mutating func fixed32(_ field: UInt32, _ value: UInt32) {
        guard value != 0 else { return }
        tag(field, Self.wireFixed32)
        bytes.append(UInt8(value & 0xff))
        bytes.append(UInt8((value >> 8) & 0xff))
        bytes.append(UInt8((value >> 16) & 0xff))
        bytes.append(UInt8((value >> 24) & 0xff))
    }

    mutating func bytesField(_ field: UInt32, _ value: [UInt8]) {
        guard !value.isEmpty else { return }
        tag(field, Self.wireLength)
        varint(UInt32(value.count))
        bytes.append(contentsOf: value)
    }

    mutating func message(_ field: UInt32, _ child: ProtoWriter) {
        bytesField(field, child.bytes)
    }
}

// MARK: - Reader

private struct ProtoField {
    let number: Int
    let wireType: Int
    /// Wire types 0 and 5 carry the value itself; wire type 2 carries the
    /// payload's byte count, with the payload in `bytes`.
    let value: UInt64
    let bytes: [UInt8]?
}

/// Walk one protobuf, tolerating unknown fields; `nil` on malformed bytes.
///
/// Tolerating the unknown is not laxity: the radio sends messages this client
/// has no model for, and it must read past them to reach the ones it needs. A
/// length that runs off the end of the buffer is a different matter and stops
/// the walk, which is what catches a truncated BLE read.
private func readFields(_ data: [UInt8]) -> [ProtoField]? {
    var out: [ProtoField] = []
    var at = 0

    func varint() -> UInt64? {
        var value: UInt64 = 0
        var shift: UInt64 = 0
        while shift < 64 {
            guard at < data.count else { return nil }
            let part = data[at]
            at += 1
            value |= UInt64(part & 0x7f) << shift
            if part & 0x80 == 0 { return value }
            shift += 7
        }
        return nil
    }

    while at < data.count {
        guard let key = varint() else { return nil }
        // Protobuf field numbers stop at 2^29-1. A key claiming more than that
        // is not a field we could ever match, and must not be truncated into
        // one that we would.
        let rawNumber = key >> 3
        let number = rawNumber > 0x1fff_ffff ? -1 : Int(rawNumber)
        let wireType = Int(key & 7)
        switch wireType {
        case ProtoWriter.wireVarint:
            guard let value = varint() else { return nil }
            out.append(ProtoField(number: number, wireType: wireType, value: value, bytes: nil))
        case ProtoWriter.wireFixed32:
            guard at + 4 <= data.count else { return nil }
            let value = UInt64(data[at])
                | UInt64(data[at + 1]) << 8
                | UInt64(data[at + 2]) << 16
                | UInt64(data[at + 3]) << 24
            at += 4
            out.append(ProtoField(number: number, wireType: wireType, value: value, bytes: nil))
        case 1:
            // Nothing this client reads is a fixed64, but one has to be stepped
            // over rather than refused, or an unrelated new field would look
            // like corruption.
            guard at + 8 <= data.count else { return nil }
            at += 8
        case ProtoWriter.wireLength:
            guard let length = varint(), length <= UInt64(data.count - at) else { return nil }
            let end = at + Int(length)
            out.append(
                ProtoField(
                    number: number,
                    wireType: wireType,
                    value: length,
                    bytes: Array(data[at..<end])
                )
            )
            at = end
        default:
            return nil
        }
    }
    return out
}

File: webapp/src/mesh/meshtasticProto.ts
/**
 * The Meshtastic client protobuf conversation, browser side.
 *
 * This is the same wire format the firmware's meshtastic_api.cpp speaks from
 * the other end: the phone (or this page) writes ToRadio, reads FromRadio
 * until config_complete echoes its nonce, then trades packets. Field numbers
 * come from meshtastic/protobufs master, and the test vectors here are the
 * same hand-computed bytes the firmware's tests pin, so the two ends cannot
 * drift apart without one of them failing.
 *
 * Pure functions over Uint8Array — no bluetooth in this file, so all of it
 * runs under node's test runner.
 */

export const MESHTASTIC_SERVICE = "6ba1b218-15a8-461f-9fa8-5dcae273eafd";
export const FROMRADIO_CHARACTERISTIC = "2c55e69e-4993-11ed-b878-0242ac120002";
export const TORADIO_CHARACTERISTIC = "f75c76d2-129e-4dad-a1dd-7866124401e7";
export const FROMNUM_CHARACTERISTIC = "ed9da18c-a800-4f66-a670-aa7547e34453";

export const BROADCAST = 0xffffffff;
const PORT_TEXT = 1;
const PORT_POSITION = 3;
const PORT_ROUTING = 5;

// ── writer ──────────────────────────────────────────────────────────────────

class Writer {
	private bytes: number[] = [];

	varint(value: number): this {
		let v = value >>> 0;
		do {
			const part = v & 0x7f;
			v >>>= 7;
			this.bytes.push(v !== 0 ? part | 0x80 : part);
		} while (v !== 0);
		return this;
	}

	tag(field: number, wire: number): this {
		return this.varint((field << 3) | wire);
	}

	uint(field: number, value: number): this {
		if (value === 0) return this;
		return this.tag(field, 0).varint(value);
	}

	fixed32(field: number, value: number): this {
		if (value === 0) return this;
		return this.fixed32Always(field, value);
	}

	fixed32Always(field: number, value: number): this {
		this.tag(field, 5);
		const v = value >>> 0;
		this.bytes.push(v & 0xff, (v >>> 8) & 0xff, (v >>> 16) & 0xff, (v >>> 24) & 0xff);
		return this;
	}

	bytesField(field: number, value: Uint8Array): this {
		if (value.length === 0) return this;
		this.tag(field, 2).varint(value.length);
		for (const b of value) this.bytes.push(b);
		return this;
	}

	message(field: number, child: Writer): this {
		return this.bytesField(field, child.finish());
	}

	finish(): Uint8Array {
		return new Uint8Array(this.bytes);
	}
}

// ── reader ──────────────────────────────────────────────────────────────────

interface Field {
	field: number;
	wire: number;
	/** wire 0/5: the numeric value; wire 2: byte length. */
	value: number;
	/** wire 2 only. */
	bytes?: Uint8Array;
}

/** Walk one protobuf, tolerating unknown fields; null on malformed bytes. */
function readFields(data: Uint8Array): Field[] | null {
	const out: Field[] = [];
	let at = 0;
	const varint = (): number | null => {
		let value = 0;
		let shift = 0;
		while (shift < 64) {
			if (at >= data.length) return null;
			const part = data[at++];
			// Beyond 32 bits JS bitwise breaks; accumulate with multiply.
			value += (part & 0x7f) * 2 ** shift;
			if ((part & 0x80) === 0) return value;
			shift += 7;
		}
		return null;
	};
	while (at < data.length) {
		const key = varint();
		if (key === null) return null;
		const field = Math.floor(key / 8);
		const wire = key % 8;
		if (wire === 0) {
			const value = varint();
			if (value === null) return null;
			out.push({ field, wire, value });
		} else if (wire === 5) {
			if (at + 4 > data.length) return null;
			const value =
				(data[at] | (data[at + 1] << 8) | (data[at + 2] << 16) | (data[at + 3] << 24)) >>> 0;
			at += 4;
			out.push({ field, wire, value });
		} else if (wire === 1) {
			if (at + 8 > data.length) return null;
			at += 8;
		} else if (wire === 2) {
			const length = varint();
			if (length === null || at + length > data.length) return null;
			out.push({ field, wire, value: length, bytes: data.subarray(at, at + length) });
			at += length;
		} else {
			return null;
		}
	}
	return out;
}

// ── ToRadio (page → deck) ───────────────────────────────────────────────────

export function encodeWantConfig(nonce: number): Uint8Array {
	const w = new Writer().tag(3, 0).varint(nonce >>> 0);
	return w.finish();
}

export function encodeTextPacket(input: {
	to: number;
	channel: number;
	packetId: number;
	text: string;
	wantAck: boolean;
}): Uint8Array {
	const data = new Writer()
		.uint(1, PORT_TEXT)
		.bytesField(2, new TextEncoder().encode(input.text));
	const packet = new Writer()
		.fixed32(2, input.to)
		.uint(3, input.channel)
		.message(4, data)
		.fixed32(6, input.packetId)
		.uint(10, input.wantAck ? 1 : 0);
	return new Writer().message(1, packet).finish();
}

// ── FromRadio (deck → page) ─────────────────────────────────────────────────

export type FromRadio =
	| { kind: "myInfo"; num: number }
	| { kind: "metadata"; firmware: string }
	| {
			kind: "nodeInfo";
			num: number;
			id: string;
			longName: string;
			shortName: string;
			snr?: number;
			lat?: number;
			lon?: number;
	  }
	| { kind: "configComplete"; nonce: number }
	| {
			kind: "text";
			from: number;
			to: number;
			packetId: number;
			channel: number;
			text: string;
			rxSnr?: number;
			rxRssi?: number;
	  }
	| { kind: "position"; from: number; lat: number; lon: number }
	| { kind: "routing"; requestId: number; error: number }
	| { kind: "other" };

const utf8 = new TextDecoder();

function toFloat(bits: number): number {
	const buffer = new ArrayBuffer(4);
	new DataView(buffer).setUint32(0, bits, true);
	return new DataView(buffer).getFloat32(0, true);
}

function toSigned32(value: number): number {
	return value | 0;
}

function parseNodeInfo(bytes: Uint8Array): FromRadio {
	const fields = readFields(bytes) ?? [];
	let num = 0;
	let id = "";
	let longName = "";
	let shortName = "";
	let snr: number | undefined;
	let lat: number | undefined;
	let lon: number | undefined;
	for (const f of fields) {
		if (f.field === 1 && f.wire === 0) num = f.value;
		if (f.field === 2 && f.bytes) {
			for (const u of readFields(f.bytes) ?? []) {
				if (!u.bytes) continue;
				if (u.field === 1) id = utf8.decode(u.bytes);
				if (u.field === 2) longName = utf8.decode(u.bytes);
				if (u.field === 3) shortName = utf8.decode(u.bytes);
			}
		}
		if (f.field === 3 && f.bytes) {
			for (const p of readFields(f.bytes) ?? []) {
				if (p.field === 1 && p.wire === 5) lat = toSigned32(p.value) / 1e7;
				if (p.field === 2 && p.wire === 5) lon = toSigned32(p.value) / 1e7;
			}
		}
		if (f.field === 4 && f.wire === 5) snr = toFloat(f.value);
	}
	return { kind: "nodeInfo", num, id, longName, shortName, snr, lat, lon };
}

function parsePacket(bytes: Uint8Array): FromRadio {
	const fields = readFields(bytes) ?? [];
	let from = 0;
	let to = 0;
	let packetId = 0;
	let channel = 0;
	let rxSnr: number | undefined;
	let rxRssi: number | undefined;
	let decoded: Uint8Array | undefined;
	for (const f of fields) {
		if (f.field === 1 && f.wire === 5) from = f.value;
		if (f.field === 2 && f.wire === 5) to = f.value;
		if (f.field === 3 && f.wire === 0) channel = f.value;
		if (f.field === 4 && f.bytes) decoded = f.bytes;
		if (f.field === 6 && f.wire === 5) packetId = f.value;
		if (f.field === 8 && f.wire === 5) rxSnr = toFloat(f.value);
		if (f.field === 12 && f.wire === 0) rxRssi = f.value > 2 ** 31 ? f.value - 2 ** 32 : f.value;
	}
	if (!decoded) return { kind: "other" };
	let portnum = 0;
	let payload: Uint8Array = new Uint8Array();
	let requestId = 0;
	for (const d of readFields(decoded) ?? []) {
		if (d.field === 1 && d.wire === 0) portnum = d.value;
		if (d.field === 2 && d.bytes) payload = d.bytes;
		if (d.field === 6 && d.wire === 5) requestId = d.value;
	}
	if (portnum === PORT_TEXT) {
		return { kind: "text", from, to, packetId, channel, text: utf8.decode(payload), rxSnr, rxRssi };
	}
	if (portnum === PORT_POSITION) {
		let lat = 0;
		let lon = 0;
		for (const p of readFields(payload) ?? []) {
			if (p.field === 1 && p.wire === 5) lat = toSigned32(p.value) / 1e7;
			if (p.field === 2 && p.wire === 5) lon = toSigned32(p.value) / 1e7;
		}
		return { kind: "position", from, lat, lon };
	}
	if (portnum === PORT_ROUTING) {
		let error = 0;
		for (const r of readFields(payload) ?? []) {
			if (r.field === 3 && r.wire === 0) error = r.value;
		}
		return { kind: "routing", requestId, error };
	}
	return { kind: "other" };
}

export function parseFromRadio(data: Uint8Array): FromRadio | null {
	const fields = readFields(data);
	if (fields === null) return null;
	for (const f of fields) {
		switch (f.field) {
			case 2:
				if (f.bytes) return parsePacket(f.bytes);
				break;
			case 3: {
				if (!f.bytes) break;
				for (const m of readFields(f.bytes) ?? []) {
					if (m.field === 1 && m.wire === 0) return { kind: "myInfo", num: m.value };
				}
				return { kind: "myInfo", num: 0 };
			}
			case 4:
				if (f.bytes) return parseNodeInfo(f.bytes);
				break;
			case 7:
				return { kind: "configComplete", nonce: f.value };
			case 13: {
				if (!f.bytes) break;
				for (const m of readFields(f.bytes) ?? []) {
					if (m.field === 1 && m.bytes) return { kind: "metadata", firmware: utf8.decode(m.bytes) };
				}
				return { kind: "metadata", firmware: "" };
			}
			default:
				break;
		}
	}
	return { kind: "other" };
}
